torch/csrc/jit/type.h

```diff
@@ -190,15 +194,21 @@ struct TupleType : public Type {
   virtual bool operator==(const Type& rhs) const override {
     if(rhs.kind() != kind())
       return false;
-    return rhs.cast<TupleType>()->elements().equals(elements());
+    const auto & l_elements = elements();
+    const auto & r_elements = rhs.cast<TupleType>()->elements();
+    if(l_elements.size() != r_elements.size())
+      return false;
+    for(size_t i = 0; i < l_elements.size(); ++i) {
+      if(*l_elements[i] != *r_elements[i])
+        return false;
+    }
+    return true;
```

Comment:
The last 5 lines can be just `std::equal(l_elements.begin(), l_elements.end(), r_elements.begin());`

Reply:
That would do `l_elements[i] = r_elements[i]`, we need `*l_elements[i] = *r_elements[i]`
It is not the same because there is a meaningful dereference in there.